File: repos/koi-web/src/__tests__/storage/user_repository.test.ts
import 'fake-indexeddb/auto';
import { db } from '../../lib/db';
import { UserRepository } from '../../lib/repositories/user_repository';
import type { User } from '../../models/index';

function makeUser(id: string): User {
  return {
    id,
    createdAt: new Date('2024-01-01'),
    lastActive: new Date('2024-01-01'),
    profile: { language: 'en' },
    settings: {},
    stats: {
      total_conversations: 0,
      total_check_ins: 0,
      current_streak: 0,
      longest_streak: 0,
      jigsaw_pieces: 0,
    },
  };
}

describe('UserRepository', () => {
  let repo: UserRepository;

  beforeEach(() => { repo = new UserRepository(); });
  afterEach(async () => { await db.users.clear(); });

  it('saves and retrieves a user', async () => {
    await repo.save(makeUser('u1'));
    const found = await repo.getById('u1');
    expect(found).toBeDefined();
    expect(found!.id).toBe('u1');
  });

  it('returns undefined for missing id', async () => {
    expect(await repo.getById('missing')).toBeUndefined();
  });

  it('updates lastActive', async () => {
    const before = new Date('2024-01-01');
    await repo.save({ ...makeUser('u2'), lastActive: before });
    await repo.updateLastActive('u2');
    const updated = await repo.getById('u2');
    expect(updated!.lastActive.getTime()).toBeGreaterThan(before.getTime());
  });

  it('deletes a user', async () => {
    await repo.save(makeUser('u3'));
    await repo.delete('u3');
    expect(await repo.getById('u3')).toBeUndefined();
  });

  it('getAll returns all users', async () => {
    await repo.save(makeUser('u4'));
    await repo.save(makeUser('u5'));
    expect((await repo.getAll()).length).toBe(2);
  });
});

File: repos/koi-backend/flutter_lib/test/storage/repositories/user_repository_test.dart
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:hive/hive.dart';
import 'package:koi_flutter_lib/models/hive_user_model.dart';
import 'package:koi_flutter_lib/storage/hive_manager.dart';
import 'package:koi_flutter_lib/storage/repositories/user_repository.dart';

HiveUser makeUser(String id) => HiveUser(
  id: id,
  createdAt: DateTime(2024, 1, 1),
  lastActive: DateTime(2024, 1, 1),
  language: 'en',
  settings: '{}',
  totalConversations: 0,
  totalCheckIns: 0,
  currentStreak: 0,
  longestStreak: 0,
  jigsawPieces: 0,
);

void main() {
  late Directory tempDir;
  late UserRepository repo;

  setUp(() async {
    tempDir = await Directory.systemTemp.createTemp('hive_test_');
    Hive.init(tempDir.path);
    if (!Hive.isAdapterRegistered(0)) Hive.registerAdapter(HiveUserAdapter());
    await Hive.openBox<HiveUser>(HiveManager.usersBox);
    repo = UserRepository();
  });

  tearDown(() async {
    await Hive.close();
    tempDir.deleteSync(recursive: true);
  });

  test('saveUser and getUser round-trip', () async {
    final user = makeUser('u1');
    await repo.saveUser(user);
    final found = await repo.getUser('u1');
    expect(found, isNotNull);
    expect(found!.id, equals('u1'));
    expect(found.language, equals('en'));
  });

  test('getUser returns null for missing id', () async {
    expect(await repo.getUser('missing'), isNull);
  });

  test('deleteUser removes entry', () async {
    final user = makeUser('u2');
    await repo.saveUser(user);
    await repo.deleteUser('u2');
    expect(await repo.getUser('u2'), isNull);
  });

  test('hasUser returns correct value', () async {
    await repo.saveUser(makeUser('u3'));
    expect(repo.hasUser('u3'), isTrue);
    expect(repo.hasUser('none'), isFalse);
  });

  test('updateLastActive changes timestamp', () async {
    final before = DateTime(2024, 1, 1);
    final user = makeUser('u4')..lastActive = before;
    await repo.saveUser(user);
    await repo.updateLastActive('u4');
    final updated = await repo.getUser('u4');
    expect(updated!.lastActive.isAfter(before), isTrue);
  });
}
